### PR TITLE
SourceLocation.init should enforce valid arguments

### PR DESCRIPTION
**Pull request by stmontgomery**
This modifies `SourceLocation.init()` to add `precondition`s which enforce that its arguments are valid, matching preconditions in the corresponding properties.

### Modifications:

- Add `precondition`s to the initializer which match those in `didSet` of the corresponding properties.
- Add precondition DocC symbol documentation where relevant.
- Adjusted existing tests which were previously invalid (although not particularly problematic) but were not failing.

### Result:

Newly-added exit tests now pass.

### Checklist:

- [x] Code and documentation should follow the style of the [Style Guide](https://github.com/apple/swift-testing/blob/main/Documentation/StyleGuide.md).
- [x] If public symbols are renamed or modified, DocC references should be updated.

### CODE REVIEW

**Comment by stmontgomery** [comment]
@swift-ci please test

File: Tests/TestingTests/SourceLocationTests.swift [hunk 1]
```diff
@@ -51,21 +51,6 @@ struct SourceLocationTests {
     #expect(sourceLocation.fileName == "D.swift")
   }
 
-
-#if !SWT_NO_EXIT_TESTS
```
Comment (stmontgomery) [inline Tests/TestingTests/SourceLocationTests.swift:55]: I just moved this down without modification, to consolidate several exit tests together

File: Sources/Testing/SourceAttribution/SourceLocation.swift [hunk 2]
```diff
@@ -12,13 +12,17 @@
...
+  ///   [`#fileID`](https://developer.apple.com/documentation/swift/fileID()).
+  ///   macro in the Swift standard library.
+  ///
   /// ## See Also
   ///
   /// - ``moduleName``
   /// - ``fileName``
   public var fileID: String {
     didSet {
-      precondition(!fileID.isEmpty)
```
Comment (grynspan) [inline Sources/Testing/SourceAttribution/SourceLocation.swift:21]: Why did you delete this precondition? While I recognize the one after it covers this scenario, the requirement that the string not be empty is explicit here.

File: Sources/Testing/SourceAttribution/SourceLocation.swift [hunk 3]
```diff
@@ -12,13 +12,17 @@
 public struct SourceLocation: Sendable {
   /// The file ID of the source file.
   ///
+  /// - Precondition: The value of this property must be non-empty and be
```
Comment (grynspan) [inline Sources/Testing/SourceAttribution/SourceLocation.swift:15]:
```suggestion
  /// - Precondition: The value of this property must not be empty and must be
```

File: Sources/Testing/SourceAttribution/SourceLocation.swift [hunk 4]
```diff
@@ -74,20 +78,48 @@ public struct SourceLocation: Sendable {
   public var _filePath: String
 
   /// The line in the source file.
+  ///
+  /// - Precondition: The value of this property must be greater than or equal
```
Comment (grynspan) [inline Sources/Testing/SourceAttribution/SourceLocation.swift:82]: You could rewrite this as just "greater than `0`".

File: Sources/Testing/SourceAttribution/SourceLocation.swift [hunk 5]
```diff
@@ -74,20 +78,48 @@ public struct SourceLocation: Sendable {
...
+  ///   to `1`.
   public var line: Int {
     didSet {
       precondition(line > 0)
     }
   }
 
   /// The column in the source file.
+  ///
+  /// - Precondition: The value of this property must be greater than or equal
```
Comment (grynspan) [inline Sources/Testing/SourceAttribution/SourceLocation.swift:92]: As above.

File: Sources/Testing/SourceAttribution/SourceLocation.swift [hunk 6]
```diff
@@ -74,20 +78,48 @@ public struct SourceLocation: Sendable {
...
+  ///     the documentation for the
+  ///     [`#fileID`](https://developer.apple.com/documentation/swift/fileID())
+  ///     macro in the Swift standard library.
+  ///   - filePath: The path to the source file.
+  ///   - line: The line in the source file. Must be greater than or equal to
+  ///     `1`.
+  ///   - column: The column in the source file. Must be greater than or equal
+  ///     to `1`.
+  ///
+  /// - Precondition: `fileID` must be non-empty and be formatted as described
```
Comment (grynspan) [inline Sources/Testing/SourceAttribution/SourceLocation.swift:113]: As above.

**Review by grynspan — CHANGES_REQUESTED** [review]
We should consider adding comments to all the `precondition()` calls since they don't do source capture, e.g.:

```swift
precondition(line > 0, "SourceLocation.line must be greater than 0 (was \(line))")
```

File: Sources/Testing/SourceAttribution/SourceLocation.swift (hunk 4 above)
Comment (stmontgomery) [reply Sources/Testing/SourceAttribution/SourceLocation.swift:82]: Sure. I had adapted the phrasing from a doc comment I saw you had written for `Configuration.maximumIterationCount` 😜

File: Sources/Testing/SourceAttribution/SourceLocation.swift (hunk 2 above)
Comment (stmontgomery) [reply Sources/Testing/SourceAttribution/SourceLocation.swift:21]: It felt redundant, but I can see how this check is more efficient at ruling out certain invalid values so I can leave it in.

**Comment by stmontgomery** [comment]
@swift-ci please test

File: Sources/Testing/SourceAttribution/SourceLocation.swift [hunk 7]
```diff
@@ -12,14 +12,19 @@
 public struct SourceLocation: Sendable {
   /// The file ID of the source file.
   ///
+  /// - Precondition: The value of this property must not be empty and be
```
Comment (grynspan) [inline Sources/Testing/SourceAttribution/SourceLocation.swift:15]:
```suggestion
  /// - Precondition: The value of this property must not be empty and must be
```

_grammarz_

File: Sources/Testing/SourceAttribution/SourceLocation.swift [hunk 8]
```diff
@@ -12,14 +12,19 @@
...
   ///
   /// - ``moduleName``
   /// - ``fileName``
   public var fileID: String {
-    didSet {
-      precondition(!fileID.isEmpty)
-      precondition(fileID.contains("/"))
+    willSet {
+      precondition(!newValue.isEmpty, "SourceLocation.fileID must not be empty (was \(newValue))")
+      precondition(newValue.contains("/"), "SourceLocation.fileID must use the format described in documentation for #fileID (was \(newValue))")
```
Comment (grynspan) [inline Sources/Testing/SourceAttribution/SourceLocation.swift:27]: Nitpick (feel free to disregard) but we can probably just say `"SourceLocation.fileID must be a well-formed file ID (was \(newValue))"`. No need to call out the docs here. (Just a thought on style.)

**Review by grynspan — APPROVED** [review]

**Comment by stmontgomery** [comment]
@swift-ci please test
